gh-127266: avoid data races when updating type slots

nascheme opened the pull request:
Avoid data races when updating type slots and type flags by "stopping-the-world" first.  This is only needed if the type is potentially exposed to multiple threads.  There are a couple places this happens:

- `type_setattro()` when the assigned item is a "dunder" method that corresponds to one or more slot
- assigning to `__bases__`
- `_abc.c` calling `_PyType_SetFlags()` to set the collection flags (SEQUENCE or MAPPING)

Two examples of code that causes this: `dataclasses` will assign dunder methods after the type object has been created, when the decorator runs.  That triggers this and since the slots are assigned one-by-one, we stop for each assignment.  Not great.   Another example is unittest `MagicMock`.  Again, it assigns dunder methods after the class is defined.

In order to assess the performance impact of this change, I've measured the following things.  The time it takes to run the unit test suite does not seem significantly worse, even when running tests serially using `-j 1`.   Comparing the pyperformance results before and after this change shows no significant change.  See link below for the chart.

Running with TSAN enabled and running the `--tsan` and `--tsan-parallel --parallel-threads=4` tests show no warnings.  I also have a modified pyperformance test suite that I can run in parallel and that shows no TSAN warnings related to this as well.

Additional changes after feedback from Sam:

- Replace the critical sections that use `TYPE_LOCK` with a simple mutex instead.  To avoid dead-locks, the code was analyzed for re-entrancy and adjusted as needed.  There are a couple places where the mutex is dropped before calling something that's possibly re-entrant.
- Remove some locking where it is not needed.  Slots can only be changed when the world is stopped (e.g. reading `tp_mro` no longer needs to hold the lock).
- Create `type_lookup_ex()` as a lower-level version of ` _PyType_LookupStackRefAndVersion` that optionally acquires the `TYPE_LOCK` or requires that it is already held.
- Add a new type flag that is used for debug enabled builds.  It tries to check if a type is potentially exposed to multiple threads (and therefore needs stop-the-world before modifying slots).   This flag could perhaps be removed before the final merge of this PR (since type flag bits are kind of in short supply).

<!-- [Benchmarks using pyperformance](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250313-3.14.0a5%2B-d4ce112-NOGIL/bm-20250313-vultr-x86_64-nascheme-gh_127266_type_slots-3.14.0a5%2B-d4ce112-vs-base.svg) -->

[Benchmark using pyperformance](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250326-3.14.0a6%2B-ca00e74-NOGIL/bm-20250326-vultr-x86_64-nascheme-gh_127266_type_slots-3.14.0a6%2B-ca00e74-vs-base.svg)


<!-- gh-issue-number: gh-127266 -->
* Issue: gh-127266
<!-- /gh-issue-number -->

nascheme committed:
wip: update type slots, stop-the-world

nascheme committed:
Remove unneeded atomics for tp_flags.

nascheme committed:
Use stop-the-world for tp_flag changes too.

nascheme committed:
Remove 'world_stops' and 'sys._get_world_stops'.

This would be better to be part of the pystats info.

nascheme committed:
Improve code comments.

nascheme committed:
Remove TSAN suppressions that seem unneeded.

nascheme committed:
Add NEWS file.

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -5779,7 +5792,17 @@ void
+    unsigned long new_flags = (self->tp_flags & ~mask) | flags;
+    if (new_flags != self->tp_flags) {
+#ifdef TYPE_FLAGS_UPDATE_NEEDS_STOP
+        PyInterpreterState *interp = _PyInterpreterState_GET();
```
if the world is stopped when assigning flags then I think acquiring type lock above would be unnecessary

colesbury replied:
Yeah, that sound right to me, but the code needs to be adjusted. We can't safely compute `new_flags` outside of the stop-the-world because the flags may change while waiting for `_PyEval_StopTheWorld()`.

Something like:

```
if ((self->tp_flags & ~mask) | flags != flags) {
  PyInterpreterState *interp = _PyInterpreterState_GET();
  _PyEval_StopTheWorld(interp);
  self->tp_flags = (self->tp_flags & ~mask) | flags
  _PyEval_StartTheWorld(interp);
}
```

nascheme replied:
Fixed as suggested.

colesbury commented on Objects/typeobject.c:
```diff
@@ -84,6 +84,17 @@ class object "PyObject *" "&PyBaseObject_Type"
+// If defined, type slot updates (after initial creation) will stop-the-world.
+#define TYPE_SLOT_UPDATE_NEEDS_STOP
+// If defined, type flag updates (after initial creation) will stop-the-world.
+#define TYPE_FLAGS_UPDATE_NEEDS_STOP
```
If we need conditional compilation, let's just stick to `#ifdef Py_GIL_DISABLED`. Having extra conditional macros complicates the maintenance because it's clear if what combinations of `TYPE_SLOT_UPDATE_NEEDS_STOP`, `TYPE_FLAGS_UPDATE_NEEDS_STOP`, and `Py_GIL_DISABLED` we intend to support.

colesbury commented on Objects/typeobject.c:
```diff
@@ -11241,20 +11299,33 @@ fixup_slot_dispatchers(PyTypeObject *type)
+    // races.  We don't use update_slot_world_stopped() here because we want
+    // to stop once rather than once per slot.
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    _PyEval_StopTheWorld(interp);
```
`_PyEval_StopTheWorld()` is always defined and is a no-op in the default build.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5972,9 +6010,15 @@ type_update_dict(PyTypeObject *type, PyDictObject *dict, PyObject *name,
 
+#ifdef TYPE_SLOT_UPDATE_NEEDS_STOP
+    if (is_dunder_name(name) && has_slotdef(name)) {
+        return update_slot_world_stopped(type, name);
```
I'm concerned that using a mix of `BEGIN_TYPE_LOCK()` and stop the world events will break atomicity of some operations. The type lock is currently a critical section and may be released while this thread is blocked on a stop the world event.

I think this means that you can end up with inconsistency between the slots and the attributes in the dictionary and not just a transient one. Consider two threads that are modifying a class's dunder method:

1) T1: update dict
2) T2: update dict
3) T2: update slot
4) T1: update slot

So once everything is finished, the dict reflect's T2's modification but the slots reflect T1's modification.

nascheme replied:
I've changed to using a regular mutex.

colesbury reviewed (commented):
I'm concerned about atomicity here (see comment below).

nascheme committed:
Use mutex rather than critical sections.

For TYPE_LOCK, replace critical sections with a simple mutex.  A number
of changes were needed to avoid deadlocking on reentrancy (trying to
re-acquire the mutex). Split _PyType_LookupRefAndVersion() into a locked
and unlocked version.  Remove atomic operations where they are not
required. Remove some cases of TYPE_LOCK being held that is not
required.

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Fix non-debug build.

nascheme committed:
Improve comments.

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Avoid unused function warning.

colesbury commented on Objects/typeobject.c:
```diff
@@ -6056,10 +6190,26 @@ _Py_type_getattro(PyObject *tp, PyObject *name)
 
+#ifdef Py_GIL_DISABLED
+static int
+update_slot_world_stopped(PyTypeObject *type, PyObject *name)
```
In other places we use the "_world_stopped" suffix to mean that the function must be called in a stop the world pause (as opposed to internally performing one).

I think we can avoid the naming ambiguity by just inlining this into the only caller.

nascheme replied:
Inlined as suggested.

colesbury commented on Objects/typeobject.c:
```diff
@@ -60,30 +60,113 @@ class object "PyObject *" "&PyBaseObject_Type"
 
 #define ASSERT_TYPE_LOCK_HELD() \
-    _Py_CRITICAL_SECTION_ASSERT_MUTEX_LOCKED(TYPE_LOCK)
+    {if (!types_world_is_stopped()) assert(types_mutex_is_owned());}
```
Can this be an expression:

```
assert(types_world_is_stopped() || types_mutex_is_owned())
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -5884,23 +6012,29 @@ _PyType_CacheGetItemForSpecialization(PyHeapTypeObject *ht, PyObject *descriptor
-    BEGIN_TYPE_LOCK();
-    type_set_flags_with_mask(self, mask, flags);
-    END_TYPE_LOCK();
+    types_mutex_lock();
```
I don't think this needs to hold the type lock. If I understand correctly, `tp_flags` is only written during a stop the world pause (or before the type is visible to other threads).

As long as all the modifications are during a stop-the-world pause, we don't need any locking for reads. (We just need an active PyThreadState).

colesbury commented on Tools/tsan/suppressions_free_threading.txt:
```diff
@@ -13,14 +13,13 @@
 
 race_top:assign_version_tag
-race_top:_Py_slot_tp_getattr_hook
+race_top:_multiprocessing_SemLock_acquire_impl
```
Did you intend to re-add `_multiprocessing_SemLock_acquire_impl` ?

nascheme replied:
Nope, that was a merge failure.  I'll fix.

colesbury commented on Include/object.h:
```diff
@@ -576,8 +576,8 @@ given type object has a specified feature.
-/* Unused. Legacy flag */
-#define Py_TPFLAGS_VALID_VERSION_TAG  (1UL << 19)
+/* Type structure is potentially exposed (revealed) to other threads */
+#define Py_TPFLAGS_EXPOSED  (1UL << 19)
```
Let's revert this before it's committed. I'd rather not change a publicly visible flag.

If needed, we can use a bit in `ob_flags`.

nascheme replied:
I think the debug asserts are useful enough to allocate a ob_flags value for that purpose.  We can remove later if we decide we need more bits.

colesbury commented on Objects/typeobject.c:
```diff
@@ -3464,7 +3549,9 @@ mro_internal_unlocked(PyTypeObject *type, int initial, PyObject **p_old_mro)
     old_mro = Py_XNewRef(lookup_tp_mro(type));
+    types_mutex_unlock();
     new_mro = mro_invoke(type);  /* might cause reentrance */
+    types_mutex_lock();
```
This doesn't seem right to me. This can be called with both the type lock held and the world stopped. If `mro_invoke` can call arbitrary code, it needs to start the world around the call.

nascheme replied:
Yes, that can call a custom `mro()` method on the meta class.  The only thing in Lib that does that is a `test_descr` test but obviously we need to handle it.

nascheme replied:
I believe this has been resolved.   Inside `mro_invoke()` I either start-the-world before calling `mro()` (if it's custom) or I unlock the mutex.  The latter case happens during `type_ready()`.

colesbury commented on Objects/typeobject.c:
```diff
@@ -1848,9 +1923,18 @@ type_set_bases(PyObject *tp, PyObject *new_bases, void *Py_UNUSED(closure))
     PyTypeObject *type = PyTypeObject_CAST(tp);
     int res;
-    BEGIN_TYPE_LOCK();
+    bool need_stop = !types_world_is_stopped();
```
This looks fishy: `type_set_bases` is called from Python code. We shouldn't be running Python code with the world stopped.

nascheme replied:
Okay, this needs reworking then.  This calls `invoke_mro()` which ends up running user code and that could re-enter.  I think a fix could be to push the stop-the-world lower down and don't be stopped when `mro()` is called.  That method is a real pain. :-(

I'll try to fix this.

nascheme replied:
I made the fix to `type_set_bases()` as suggested.  For `mro_invoke()` we need to decide if we are stopped or if the mutex is held.

colesbury reviewed (commented):
Nice!

I think there's still too much mixing of the type lock and the stop the world. I think data should either be protected by the type lock OR by stop-the-world, but not both.

From reading through this:

type lock:
* tp_version_tag
* _spec_cache

stop the world:
* tp_flags (modifications)
* bases (modifications)

Additionally, there's a few functions that are named with `_lock_held` but are called in a stop-the-world pause (not with the type lock held), which I found a bit confusing.

nascheme committed:
Remove unwanted suppression (bad merge).

nascheme committed:
Fixes based on review feedback.

* In type_set_bases(), always use stop-the-world.

* Fixes for mro_invoke().  We need to do different things depending
  on if there is a custom mro() method and where we are called from.
  If from type_ready(), need the TYPE_LOCK mutex.  If from assignment
  to __bases__, need stop-the-world.

* Remove some places TYPE_LOCK is acquired when it no longer needs
  to be.

* Add comments to some _lock_held() functions to note that they might
  be called with the world stopped, rather than holding the lock.

* Remove update_slot_world_stopped() by inlining it.

nascheme committed:
Remove spurious assert().

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

kumaraditya303 commented on Include/internal/pycore_interp_structs.h:
```diff
@@ -566,6 +566,8 @@ struct types_state {
     } for_extensions;
     PyMutex mutex;
+    // used to check correct usage of the above mutex
+    unsigned long long mutex_tid;
```
This is used only in free-threading builds right? I think we should guard it with `ifdef Py_GIL_DISABLED`.

nascheme replied:
Done.

nascheme committed:
Omit mutex_tid member from default build.

nascheme committed:
Remove Py_TPFLAGS_EXPOSED flag and related logic.

This was useful for debugging logic but we don't want to use an
additional type flag.

nascheme committed:
Improve comment for TYPE_LOCK.

nascheme committed:
Re-add the ASSERT_NEW_OR_STOPPED() asserts.

These seem like helpful asserts that could catch future bugs, e.g. if the
typeobject.c code is refactored or significantly changed.  So, I think it
makes sense to allocate a ob_flags bit for this purpose.  We can remove it
later if desired since it's non-public.

Move the ob_flags bit values into the object.h header file.

nascheme committed:
Further cleanups of the locking in typeobject.

Reduce the number of places and the regions where TYPE_LOCK is being
held.  Most significantly, don't take the lock inside type_ready().

This removes the need for some _lock_held variants of some functions.
We don't need type_lookup_lock_held() anymore, for example.

Rename type_set_bases_unlocked() to type_set_bases_world_stopped().

nascheme committed:
Fix data race in resolve_slotdups().

The type_slots_pname and type_slots_ptrs are interpreter globals. Rather
than trying to make access safe, just disable those caches in the
free-threaded build.  This cache is of dubious value anyhow.

nascheme committed:
Fix comment.

colesbury commented on Objects/typeobject.c:
```diff
@@ -3066,11 +3150,16 @@ tail_contains(PyObject *tuple, Py_ssize_t whence, PyObject *o)
+#ifdef Py_GIL_DISABLED
+    // Avoid re-entrancy and use tp_name
+    return type_name(cls, NULL);
+#else
```
This seems fishy

nascheme replied:
Maybe a little.  :-)   The return value is only used for the type name in the error message about having duplicate classes in the MRO.  I think using `tp_name` for that error is okay.

Keeping the logic that effectively does `return cls.__name__` is more complicated.   We would have to start-the-world, if stopped and then re-stop it again after the getattr call.

colesbury reviewed (commented):
I'm still concerned about the locking discipline and think we need to further refactor these functions. The `mro_invoke`, `mro_implementation`, `mro_hierarchy` functions are particularly messy with respect to locking and stop the world.

For `type_setattro`/`type_update_dict`, it seems like we can just determine ahead of time if we need to lock or stop the world based on the attribute name.

nascheme committed:
Make the init of tp_dict thread-safe.

It is unlikely that tp_dict is not allocated when type_setattro() is
called.   However, if it is NULL, stop-the-world so we can safely assign
it.

nascheme committed:
Do some addtional locking simplification.

Re-structure type_setattro() into a slot and non-slot case.  For the
slot case, we need to stop-the-world.  For non-slot, we need to hold
TYPE_LOCK.

Remove unneeded locking in type_mro_modified().  This is only called if
the type is new or if the world is stopped.  In both cases, TYPE_LOCK
doesn't need to be held.

sergey-miryanov commented on Objects/typeobject.c:
```diff
@@ -10995,12 +11084,21 @@ resolve_slotdups(PyTypeObject *type, PyObject *name)
 {
     /* XXX Maybe this could be optimized more -- but is it worth it? */
 
+#ifdef Py_GIL_DISABLED
```
JFYI, I have removed resolve_slotdups here - gh-132156

sergey-miryanov commented on Objects/typeobject.c:
```diff
@@ -11025,11 +11125,25 @@ resolve_slotdups(PyTypeObject *type, PyObject *name)
+// Return true if "name" corresponds to at least one slot definition.  This is
+// a more accurate but more expensive test compared to is_dunder_name().
+static bool
+has_slotdef(PyObject *name)
```
Also, I built a temp dictionary on main interp init that can be used for this. In final version I didn't store it in interp state but maybe it is worth to store it (I have intermediate commits for this - https://github.com/python/cpython/pull/132156/commits/bac95a56806a5a9de5eb5ce3cf7dbd1520662cbe, https://github.com/python/cpython/pull/132156/commits/b0ad8754b45a79fe5b45063363384b9c236228d3, https://github.com/python/cpython/pull/132156/commits/79a165d4a6998f33e29c908b424ccb862fd191eb)
JFYI.

nascheme committed:
Avoid acquiring the types mutex if version is set.

This can potentially help quite a bit when the type version tag has
already been assigned but the MCACHE is missed.  We can complete
the lookup without acquiring the types mutex.

nascheme committed:
Add check_invalid_reentrancy() call.

nascheme committed:
Fix additional re-entrancy issues found.

- check_set_special_type_attr() is potentially re-entrant.  Fix
  by refactoring type_set_bases_world_stopped() into two functions,
  one to check the args and one to actually set tp_base.

- the type watcher stuff is re-entrant.  Fix by unlocking the mutex
  before handling those.

- add `is_new_type` parameter to mro_internal() and friends.  This
  explicitly tells us if we are initializing a new type or modifying the
  bases of an existing one.  The stop-the-world mechanism is used in
  only the second case.

- rename best_base() to find_best_base()

nascheme committed:
Add comment explaining class_name() code.

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Use atomic load to avoid thread safety issue.

nascheme committed:
Fix default debug build.

nascheme committed:
Move declaration to avoid syntax error.

colesbury commented on Objects/typeobject.c:
```diff
@@ -1248,7 +1322,11 @@ _PyType_LookupByVersion(unsigned int version)
 _PyType_GetVersionForCurrentState(PyTypeObject *tp)
 {
+#ifdef Py_GIL_DISABLED
+    return _Py_atomic_load_uint32_acquire(&tp->tp_version_tag);
```
This function is private and not used anywhere. We should probably just delete it.

nascheme replied:
Okay, done.

colesbury commented on Objects/typeobject.c:
```diff
@@ -1178,14 +1243,23 @@ type_mro_modified(PyTypeObject *type, PyObject *bases)
+        // Raced with another thread mutating the metaclass, possible that mro()
+        // method was assigned.
+        goto clear;
+    }
```
Why is this check added? And why is it free threading specific? There's an assertion that this function is called with the world stopped.

If `has_custom_mro()` is possibly reentrant, then the stop-the-world isn't safe and also these checks should not be free threading specific, but I think we may just want to assume that `has_custom_mro()` is not reentrant.

nascheme replied:
It seems `has_custom_mro()` is reentrant.  I've changed it to re-start the world if stopped.  This means that if the lookup for `mro()` changes the meta class as a side-effect, we take the "clear" path, disabling caching.  I think that's safe although pessimistic since it might still be okay to enable caching.   This seems like such a rare case that I think not caching is fine here.

It would not be safe to use the return value of `has_custom_mro()` in that case since it might return true but the side-effect actually created a custom `mro()` hook.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5732,23 +5842,24 @@ _PyTypes_AfterFork(void)
-    }
-    if (PyStackRef_IsNull(out)) {
-        return NULL;
+    unsigned int version = FT_ATOMIC_LOAD_UINT32_RELAXED((type)->tp_version_tag);
```
I think this predates this PR, but `tp_version_tag` is an `unsigned int` so we should use `FT_ATOMIC_LOAD_UINT_RELAXED` or similar macros/functions.

nascheme replied:
Fixed.

colesbury commented on Tools/tsan/suppressions_free_threading.txt:
```diff
@@ -13,14 +13,12 @@
 # These warnings trigger directly in a CPython function.
 
 race_top:assign_version_tag
```
Does this fix races involving `assign_version_tag`?

nascheme replied:
Yes, I've removed that suppression.

colesbury commented on Objects/typeobject.c:
```diff
@@ -6166,23 +6286,45 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+    if (is_dunder_name(name) && has_slotdef(name)) {
+        // The name corresponds to a type slot.
+        types_stop_world();
+        Py_BEGIN_CRITICAL_SECTION(dict);
```
I don't think we should call `Py_BEGIN_CRITICAL_SECTION()` within a stop-the-world pause. I think it's ineffective at best and may possibly lead to deadlocks (not 100% sure).

nascheme replied:
If that's true then we need to be extremely careful whatever APIs we call while the world is stopped.  Anything that uses a critical section internally could have the same issue, right?

The critical section on the dict is here because internally we call dict APIs that assert the critical section mutex is held.  We would need to create new APIs that don't assert that.  Seems messy to me.  I don't see why this shouldn't be okay.  When the world is stopped, all mutexes part of critical sections should be released.  So this should never block or do anything funny.   Or am I missing something?

colesbury commented on Objects/typeobject.c:
```diff
@@ -60,30 +60,114 @@ class object "PyObject *" "&PyBaseObject_Type"
+{
+    assert(!types_mutex_is_owned());
+    PyMutex_Lock(TYPE_LOCK);
+    types_mutex_set_owned(PyThread_get_thread_ident_ex());
```
If this is limited to the free-threading build, I'd be a more comfortable using `Py_ThreadId()` for this and I think it'll be faster.

nascheme replied:
I changed the code but then I removed it after switching back to a critical section.  I think `_Py_ThreadId()` would work fine and be faster, as you say.

colesbury commented on Objects/typeobject.c:
```diff
@@ -1100,6 +1160,9 @@ type_modified_unlocked(PyTypeObject *type)
     if (type->tp_watched) {
+        // We must unlock here because at least the PyErr_FormatUnraisable
+        // call is re-entrant and the watcher callback might be too.
+        types_mutex_unlock();
```
This looks like it breaks atomicity of `type_modified_unlocked()`, which doesn't seem great.

nascheme replied:
This unlock is gone now that I reverted to using a critical section for `TYPE_LOCK`.  I think the atomicity can still be broken.  If the watcher callback, for example, blocks on a mutex then the lock associated with the critical section is dropped just like my code was doing by explicitly dropping the mutex.  I suppose that's more unlikely to happen with the CS but to me that just seems like a harder to trigger bug.

I don't think the broken atomicity is actually a problem in this particular case.  After this watcher logic, we set type version tag to zero and invalidate the spec cache.  I think that's safe to do no matter what the callbacks or `PyErr_FormatUnraisable` actually did.

Hmm, I guess there is a problem if those things do lookups on this type.  In that case, the old cached values might be used when they shouldn't.  So maybe we should set the type version to zero *before* calling the watcher callbacks.  I'm not familiar with the watcher stuff so I don't know what it expects.  My guess is calling it after invalidating the cache would be better.

colesbury reviewed (commented):
I'm not convinced that the switch of the type lock from a critical section to a plain mutex is profitable here. It seems to lead to extra unlocking that breaks atomicity in places that would almost certainly be atomic before (for example, watcher callbacks, `PyErr_FormatUnraisable`). I find it also harder to reason about the locking of the type lock in combination with a dictionary.

nascheme committed:
Remove _PyType_GetVersionForCurrentState, unused.

nascheme committed:
Fix for possible re-entrancy in has_custom_mro().

nascheme committed:
Use correct FT_ATOMIC_ macro.

nascheme committed:
Remove TSAN suppression for 'assign_version_tag'.

nascheme committed:
Small efficiency fix for types_mutex_set_owned().

nascheme committed:
Revert to using critical section with TYPE_LOCK.

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Invalidate type cache before calling watchers.

This avoids the watcher callback potentially using invalid cache
entries.

nascheme committed:
Fixes for type_modified_unlocked().

* Since type_modified_unlocked() can be called with the world stopped
  then we need to re-start if making re-entrant calls (like calling
  type watcher callbacks).

* Re-order code before type_modified_unlocked().  This fixes a potential
  bug if called functions end up re-assigning the type version tag.  If
  they do, the cache can contain invalid entries.  We should do the
  dictionary update first then set the version to zero.

* Replace 'is_new_type' with 'world_stopped' boolean parameter.  This
  seems a bit more clear.

nascheme committed:
Major re-work, TYPE_LOCK protects more things.

Rather than trying to update tp_base, tp_bases, and tp_mro while the
world is stopped, hold TYPE_LOCK when updating them.  This is much more
similar to how the base version of the code works and requires many
fewer changes.  Using stop-the-world turned out to be difficult to make
work correctly, mostly due to custom mro() methods but also for some
other reasons (for example, functions that potentially re-enter the
interpreter).

This re-work requires that we can call stop-the-world inside a critical
section and have the mutex still held after re-starting.  This seems to
work and I've added asserts to confirm it.  Stop-the-world is now only
used when updating either tp_flags or other type slots and not the three
type members listed above.

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Fix non-debug build.

nascheme committed:
Revert unneeded code changes.

nascheme committed:
Merge branch 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Restore comment

nascheme committed:
Revert more changes.

These are not required as part of this PR and can be reviewed
separately.

colesbury reviewed (approved):
This LGTM once the issue with `test_opcache` is resolved

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Reduce item list size for a few tests.

Now that these slot updates use stop-the-world, these two tests are
quite a lot slower.  Reduce size of the items list from 1000 to 100.

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -1708,9 +1765,9 @@ mro_hierarchy(PyTypeObject *type, PyObject *temp)
+type_check_new_bases(PyTypeObject *type, PyObject *new_bases, PyTypeObject **best_base)
 {
-    // Check arguments
+    // Check arguments, this re-entrant due to the PySys_Audit() call
```
```suggestion
    // Check arguments, this is re-entrant due to the PySys_Audit() call
```

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -74,8 +82,59 @@ class object "PyObject *" "&PyBaseObject_Type"
+    if (!types_world_is_stopped()) { _Py_CRITICAL_SECTION_ASSERT_MUTEX_LOCKED(TYPE_LOCK); }
+
+// Checks if we can safely update type slots or tp_flags.
+#define ASSERT_NEW_OR_STOPPED(tp) \
```
```suggestion
#define ASSERT_WORLD_STOPPED_OR_NEW_TYPE(tp) \
```

I think it is better to have a more descriptive name similar to `ASSERT_WORLD_STOPPED_OR_DICT_LOCKED` from `dictobject.c`

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -1586,10 +1638,13 @@ type_set_abstractmethods(PyObject *tp, PyObject *value, void *Py_UNUSED(closure)
+    types_stop_world();
     if (abstract)
-        type_add_flags(type, Py_TPFLAGS_IS_ABSTRACT);
+        type_set_flags_with_mask(type, 0, Py_TPFLAGS_IS_ABSTRACT);
```
I think this doesn't need to be changed, it can still be `type_add_flags` as mask=0

nascheme committed:
Merge 'origin/main' into gh-127266-type-slots-ts

nascheme committed:
Minor code tidy.
